src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java

```diff
@@ -93,6 +104,79 @@ public boolean isAdvanced() {
     }
 
     /**
+     * Some steps, such as {@code CoreStep} or {@code GenericSCMStep} can take
+     * arbitrary {@link Describable}s of a certain type and execute it as a step.
+     * Such a step should return true from this method so that {@link Describable}s that
+     * it supports can be directly written as a step as a short-hand.
+     *
+     * <p>
+     * Meta-step works as an invisible adapter that creates an illusion that {@link Describable}s are
+     * steps.
+     *
+     * <p>
+     * For example, in Jenkins Pipeline, if there is a meta step that can handle a {@link Describable},
+     * and it has a symbol, it allows the following short-hand:
+     *
+     * <pre>
+     * public class Xyz extends Foo {
+     *     &#64;DataBoundConstructor
+     *     public Xyz(String value) { ... }
+     *
+     *     &#64;Extension &#64;Symbol("xyz")
+     *     public static class DescriptorImpl extends FooDescriptor { ... }
+     * }
+     *
+     * public class MetaStepForFoo extends AbstractStepImpl {
+     *     &#64;DataBoundConstructor
+     *     public MetaStepForFoo(Foo delegate) {
+     *         ...
+     *     }
+     *
+     *     ...
+     *     &#64;Extension
+     *     public static class DescriptorImpl extends AbstractStepDescriptorImpl {
+     *         &#64;Override
+     *         public String getFunctionName() {
+     *             return "metaStepForFoo";
+     *         }
+     *     }
+     * }
+     *
+     * // this is the short-hand that users will use
+     * xyz('hello')
+     * // but this is how it actually gets executed
+     * metaStepForFoo(xyz('hello'))
+     * </pre>
+     *
+     * <p>
+     * Meta-step must have a {@link DataBoundConstructor} whose first argument represents a
+     * {@link Describable} that it handles.
+     */
+    public boolean isMetaStep() {
+        return false;
+    }
+
+    /**
+     * For a {@linkplain #isMetaStep() meta step}, return the type that this meta step handles.
+     * Otherwise null.
+     */
+    // not @CheckForNull because often the caller is using allMeta()
```

Comment:
I don't think adding `@Nullable` to return value to anything as far as tools are concerned.